### DISCUSSION TITLE
Unloading assembly with Avalonia controls

### DISCUSSION THREAD

**Discussion by MaxwellDAssistek** [AvaloniaUI/Avalonia#17100] [Q&A]
Hello, 

I'm working on an application where we are loading and unloading plugins with Avalonia controls in them. Each plugin is loaded into its own AssemblyLoadContext. The problem seems that those modules seem to remain loaded because the types used in them are still registered in Avalonia in some way.

One of the places that I found is AvaloniaPropertyRegistry. I did try the new UnregisterByModule method, but if I inspect the internal _properties dictionary using a debugger after calling that method I can still see various properties remain registered.

This is what I have set up in my AssemblyLoadContext currently:

```
        Unloading += context =>
        {
            foreach (var contextAssembly in context.Assemblies)
            {
                AvaloniaPropertyRegistry.Instance.UnregisterByModule(contextAssembly.DefinedTypes);
                var name = contextAssembly.GetName().Name;
                if (name != null)
                {
                    AssetLoader.InvalidateAssemblyCache(name);
                }
            }
        };
```

**Comment by stevemonaco** [comment]
Not familiar with this area, but check this sample if you haven't:

https://github.com/AvaloniaUI/Avalonia/blob/2299d96f9851aaf9b3ccc5390d69b6e8d6f877a0/samples/UnloadableAssemblyLoadContext/UnloadableAssemblyLoadContext/AssemblyLoadContextH.cs#L22-L31

You should also remove any added styles and resources. The sample doesn't seem to remove resources though, but maybe it's just not obvious. The plugin doesn't have any resources because it opted to use a `Style` instead of a `ControlTheme`. 🤔

**Comment by dhhunter** [comment]
@MaxwellDAssistek I have been hacking together an integrated Module builder for my demo Avalonia Prism application and I have a similar problem in that the Prism Module loader does not actually support unloading of modules...  My integrated module builder is such an awesome exciting feature that is really just a big memory leak right now lol.  I am not yet ready to face down that mountain of a task that is trying to figure out how to make an Unloadable Prism Module loader but thank you for asking this question because it will no doubt help me in the future!

**Comment by MaxwellDAssistek** [comment]
I am kind of stumped at this point. I was able to eventually eliminate enough things like DialogHost and now the backing class behind the view in the module is getting cleaned up, but the Xaml object is still getting left behind.

![image](https://github.com/user-attachments/assets/7ed5be71-95ec-486f-9fdb-721fa1c261b4)
